### PR TITLE
Add subcodes to Protocols

### PR DESCRIPTION
**Pull request by Yoann-Abbes**
## What does this PR do ?

Jira : [Story](https://jira.kaliop.net/browse/KZL-1312)

Add subcodes to Protocols

Add/adapt unit test.

### How should this be manually tested?

unit/functional tests

### Boyscout

80 columns
add tag "deprecated" for IndexCache

### CODE REVIEW

File: lib/api/core/entrypoints/embedded/protocols/protocol.js [hunk 1]
```diff
@@ -105,6 +107,29 @@ class Protocol {
...
+
+  /**
+   * Manage errors to be thrown
+   *
+   * @param {...args} ...args
+   * 
+   */
+
+  throw(...args) {
+
```
Comment (alexandrebouthinon) [inline lib/api/core/entrypoints/embedded/protocols/protocol.js]:
```suggestion
```

Comment (alexandrebouthinon) [inline lib/api/core/entrypoints/embedded/protocols/protocol.js]:
```suggestion
```

File: test/api/core/entrypoints/embedded/protocols/http.test.js [hunk 2]
```diff
@@ -71,6 +72,7 @@ describe('/lib/api/core/entrypoints/embedded/protocols/http', () => {
     };
 
     protocol = new HttpProtocol();
+
```
Comment (alexandrebouthinon) [inline test/api/core/entrypoints/embedded/protocols/http.test.js]:
```suggestion
```

File: test/util/errorMatcher.js [hunk 3]
```diff
@@ -1,20 +1,20 @@
 const
+  errorsManager = require('../../lib/config/error-codes/throw'),
   stableStringify = require('json-stable-stringify'),
   {
-    Request,
-    errors
+    Request
```
Comment (alexandrebouthinon) [inline test/util/errorMatcher.js]: Maybe inline this require

File: lib/api/core/entrypoints/embedded/protocols/http.js [hunk 4]
```diff
@@ -394,7 +393,7 @@ class HttpProtocol extends Protocol {
         streamLength += chunk.length;
 
         if (streamLength > maxRequestSize) {
-          callback(new SizeLimitError('Maximum HTTP request size exceeded'));
+          callback(errorsManager.getError('protocols', 'http', 'http_request_size_exceeded'));
```
Comment (scottinet) [inline lib/api/core/entrypoints/embedded/protocols/http.js]:
```suggestion
          callback(this.getError('http_request_size_exceeded'));
```

File: lib/api/core/entrypoints/embedded/protocols/websocket.js [hunk 5]
```diff
@@ -225,7 +225,7 @@ class WebSocketProtocol extends Protocol {
         {},
         {
           connection,
-          error: new BadRequestError(e.message)
+          error: this.getError('websocket_request_error', e.message)//new BadRequestError(e.message)
```
Comment (scottinet) [inline lib/api/core/entrypoints/embedded/protocols/websocket.js]: Remove that comment.

File: lib/config/error-codes/protocols.json [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      "code": 3,
+      "errors": {
+        "websocket_request_error": {
+          "code": 1,
+          "message": "%s.",
+          "class": "BadRequestError"
+        }
+      }
+    },
+    "socketio": {
```
Comment (scottinet) [inline lib/config/error-codes/protocols.json]: I didn't see that one being used.

File: lib/api/core/entrypoints/embedded/protocols/http.js (hunk 4 above)
Comment (Yoann-Abbes) [reply lib/api/core/entrypoints/embedded/protocols/http.js]: cannot acess `this` in this function

File: lib/api/core/entrypoints/embedded/protocols/protocol.js [hunk 7]
```diff
@@ -105,6 +107,28 @@ class Protocol {
...
   }
+
+  /**
+   * Manage errors to be thrown
+   *
+   * @param {...args} ...args
+   * 
+   */
+
+  throw(...args) {
```
[the remark this answers is not in the document]
Comment (Yoann-Abbes) [reply lib/api/core/entrypoints/embedded/protocols/protocol.js]: I totally agree, nice! thanks

[the remark this answers is not in the document]
Comment (Yoann-Abbes) [reply lib/api/core/entrypoints/embedded/protocols/protocol.js]: When this PR will be merged, i'll do the change in the others branches

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [throw.js#](https://github.com/kuzzleio/kuzzle/blob/bf3b1811b9b9093a241811955d4b0bdd9192843b/lib/config/error-codes/throw.js#): Add a new line at the end of this file. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AMissingNewlineAtEndOfFile)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [throw.js#](https://github.com/kuzzleio/kuzzle/blob/39b03dd207075f1e186ec8642a9df2a7922be91d/lib/config/error-codes/throw.js#): Add a new line at the end of this file. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AMissingNewlineAtEndOfFile)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [throw.js#](https://github.com/kuzzleio/kuzzle/blob/14180ec93c3b9ae7a3654c7669fa28e4e4566a6e/lib/config/error-codes/throw.js#): Add a new line at the end of this file. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AMissingNewlineAtEndOfFile)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**